template/template.go

```diff
@@ -110,6 +152,116 @@ func (t *Template) Vars() []string {
 	return vars
 }
 
+type Values struct {
+	Messages []api.Message
+}
+
+func (t *Template) Execute(w io.Writer, v Values) error {
+	system, collated := collate(v.Messages)
+	if slices.Contains(t.Vars(), "messages") {
+		return t.Template.Execute(w, map[string]any{
+			"System":   system,
+			"Messages": collated,
+		})
+	}
+
+	var b bytes.Buffer
+	var prompt, response string
+	for i, m := range collated {
+		if m.Role == "user" {
+			prompt = m.Content
+		} else {
+			response = m.Content
+		}
+
+		if i != len(collated)-1 && prompt != "" && response != "" {
+			if err := t.Template.Execute(&b, map[string]any{
+				"System":   "",
+				"Prompt":   prompt,
+				"Response": response,
+			}); err != nil {
+				return err
+			}
+
+			prompt = ""
+			response = ""
+		}
+	}
+
+	var cut bool
+	tree := t.Template.Copy()
+	// for the last message, cut everything after "{{ .Response }}"
+	tree.Root.Nodes = slices.DeleteFunc(tree.Root.Nodes, func(n parse.Node) bool {
+		if slices.Contains(parseNode(n), "Response") {
+			cut = true
+		}
+
+		return cut
+	})
+
+	if err := template.Must(template.New("").AddParseTree("", tree)).Execute(&b, map[string]any{
+		"System": system,
+		"Prompt": prompt,
+	}); err != nil {
+		return err
+	}
+
+	_, err := io.Copy(w, &b)
+	return err
+}
+
+type messages []*api.Message
+
+// Last returns true if the message is the last message of the given role
+func (s messages) Last(role string, m *api.Message) bool {
```

Comment:
Took me a bit to understand what this function did – do we need the `role` field here?

Reply:
Or could we look at the current message's role (not sure if we need to cross-check in any cases?)

Reply:
> could we look at the current message's role

that was the original implementation but it seemed confusing since this isn't the last message but the last message of a particular role

Reply:
The part that made it hard was in the example Mistral template. 

```
[INST] {{ if and ($.Messages.Last "user" .) $.System }}{{ $.System }}{{ printf "\n\n" }}
{{- end }}
```

I'm most worried about adding a new helper function (or in this case, a method on `.Messages` as they will be forever. While it's not as pretty, could we use standard template functions?

```
{{- range $index, $message := .Messages }}
  {{- if eq $message.Role "user" }}
    {{- if and (eq $index (sub (len $.Messages) 1)) $.Tools }}
      [AVAILABLE_TOOLS] {{ toJson $.Tools }}[/AVAILABLE_TOOLS]
    {{- end }}
    [INST]
    {{- if and (eq $index (sub (len $.Messages) 1)) $.System }}
      {{ $.System }}
    {{- end }}
    {{ $message.Content }}[/INST]
  {{- else if eq $message.Role "assistant" }}
    {{- if $message.Content }}
      {{ $message.Content }}</s>
    {{- else if $message.ToolCalls }}
      [TOOL_CALLS] [
      {{- range $message.ToolCalls }}
        {{ "{" }}"name": "{{ .Function.Name }}", "arguments": {{ toJson .Function.Arguments }}{{ "}" }}
      {{- end }}]</s>
    {{- end }}
  {{- else if eq $message.Role "tool" }}
    [TOOL_RESULTS] {{ toJson $message }}[/TOOL_RESULTS]
  {{- end }}
{{- end }}
```